src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -122,10 +123,10 @@ public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listen
                 }
             }
             scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
-            if (changelogFile != null && changelogFile.length() == 0) {
-                // JENKINS-57918/JENKINS-59560: Some SCM plugins don't write anything to the changelog file in some
+            if (changelogOriginalModifiedDate != null && changelogFile.lastModified() == changelogOriginalModifiedDate) {
```

Comment:
Ok, sorry, this is on me, but I just realized that checking the last modified date alone might have some issues. The resolution of the value returned by `File.lastModified` depends on the filesystem in use, and it is pretty normal for it to only have about one second resolution, so if the checkout completes and the changelog file is written in less than one second, sometimes the modification date will be unchanged and the file will be deleted, oops!

We can fix this by checking both the old and new conditions, so that we only delete the file if it's empty _and_ the last modified date is unmodified:

```suggestion
            if (changelogFile != null && changelogFile.length() == 0
                    && changelogOriginalModifiedDate != null && changelogFile.lastModified() == changelogOriginalModifiedDate) {
```

Here is the test that I used to verify the problem locally. Without the above patch, it is flaky on my machine.

```
    @Test public void gitChangelogSmokes() {
        rr.then(r -> {
            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
            p.setDefinition(new CpsFlowDefinition(
                    "node() {\n" +
                    "  checkout(scm: [\n" +
                    "    $class: 'GitSCM',\n" +
                    "    branches: [[name: '*/master']],\n" +
                    "    userRemoteConfigs: [[url: '" + sampleGitRepo.fileUrl() + "']]\n" +
                    "  ])\n" +
                    "}", true));
            // Run first build. Since there is no baseline, there are no culprits.
            sampleGitRepo.init(); // Default user is gits@mplerepo
            sampleGitRepo.write("foo", "bar");
            sampleGitRepo.git("add", "foo");
            sampleGitRepo.git("commit", "-m", "Initial commit");
            WorkflowRun b1 = r.buildAndAssertSuccess(p);
            assertThat(b1.getCulpritIds(), equalTo(Collections.emptySet()));
            // Run second build. There is a baseline now, so we should have a culprit.
            sampleGitRepo.write("foo", "bar1");
            sampleGitRepo.git("add", "foo");
            sampleGitRepo.git("commit", "-m", "Second commit");
            WorkflowRun b2 = r.buildAndAssertSuccess(p);
            assertThat(b2.getCulpritIds(), equalTo(Collections.singleton("gits")));
        });
    }
```

Reply:
Well caught. I've got a 100% failure rate using that test and no other changes when I run it on its own. Occasional passes via `mvn test` and its retries. One had a 50% passrate, the second had a 25% passrate. I'll commit your change, add this test, re-test with workflow-job and the PCT, and push again.